File: services/schemeMapper.js
function firstValue(...values) {
  return values.find((value) => value !== undefined && value !== null && value !== "");
}

function textFrom(value) {
  if (!value) return "";
  if (typeof value === "string") return value;
  if (Array.isArray(value)) return value.map(textFrom).filter(Boolean).join("; ");
  if (typeof value === "object") {
    return firstValue(value.en, value.text, value.value, value.name, value.label, JSON.stringify(value));
  }
  return String(value);
}

function findSchemeArray(raw) {
  if (Array.isArray(raw)) return raw;
  if (!raw || typeof raw !== "object") return [];

  const candidates = [
    raw.data,
    raw.schemes,
    raw.results,
    raw.response,
    raw.data?.schemes,
    raw.data?.results,
    raw.payload?.schemes
  ];

  return candidates.find(Array.isArray) || [];
}

function schemeSourceUrl(scheme) {
  const title = textFrom(firstValue(scheme.scheme_name, scheme.schemeName, scheme.name, scheme.title)).toLowerCase();
  const category = textFrom(firstValue(scheme.category, scheme.schemeCategory, scheme.categoryName)).toLowerCase();
  const state = textFrom(firstValue(scheme.state, scheme.stateName, scheme.schemeState)).toLowerCase();

  if (title.includes("pmeg") || title.includes("employment generation")) return "https://www.msme.gov.in/";
  if (title.includes("mudra")) return "https://www.mudra.org.in/";
  if (title.includes("pradhan mantri kisan") || title.includes("pm kisan") || title.includes("kisan") || category.includes("agriculture")) return "https://pmkisan.gov.in/";
  if (title.includes("e-shram") || title.includes("shram") || category.includes("labour welfare")) return "https://eshram.gov.in/";
  if (title.includes("kaushal vikas") || category.includes("skill development") || title.includes("skill")) return "https://www.pmkvyofficial.org/";
  if (category.includes("scholarship") || title.includes("scholarship")) return "https://scholarships.gov.in/";
  if (title.includes("fasal bima")) return "https://pmfby.gov.in/";
  if (title.includes("shram yogi")) return "https://pmsym.gov.in/";
  if (title.includes("rythu bandhu") || (state.includes("telangana") && category.includes("agriculture"))) return "https://rythubandhu.telangana.gov.in/";

  return "";
}

export function mapScheme(scheme) {
  const name = firstValue(scheme.scheme_name, scheme.schemeName, scheme.name, scheme.title);
  const description = firstValue(scheme.description, scheme.schemeDescription, scheme.briefDescription, scheme.summary);
  const benefits = firstValue(scheme.benefits, scheme.benefit, scheme.benefitDetails);
  const eligibility = firstValue(scheme.eligibility, scheme.eligibilityCriteria, scheme.eligibility_details);
  const category = firstValue(scheme.category, scheme.schemeCategory, scheme.categoryName, scheme.nodalMinistryName);
  const level = firstValue(scheme.level, scheme.schemeLevel, scheme.schemeType);
  const sourceUrl = firstValue(scheme.source_url, scheme.sourceUrl, scheme.url, scheme.link) || schemeSourceUrl(scheme);

  return {
    id: firstValue(scheme.id, scheme.schemeId, scheme.slug, name),
    title: textFrom(name) || "Untitled scheme",
    description: textFrom(description),
    benefits: textFrom(benefits),
    eligibility: textFrom(eligibility),
    category: textFrom(category),
    level: textFrom(level),
    sourceUrl: sourceUrl,
    raw: scheme
  };
}

export function mapSchemes(raw) {
  return findSchemeArray(raw).map(mapScheme);
}

export function getCategories(schemes) {
  return [...new Set(schemes.map((scheme) => scheme.category).filter(Boolean))].sort();
}

export function filterSchemes(schemes, { search = "", category = "" } = {}) {
  const query = search.trim().toLowerCase();

  return schemes.filter((scheme) => {
    const matchesCategory = !category || scheme.category === category;
    const haystack = [scheme.title, scheme.description, scheme.benefits, scheme.eligibility, scheme.category]
      .join(" ")
      .toLowerCase();
    const matchesSearch = !query || haystack.includes(query);
    return matchesCategory && matchesSearch;
  });
}

File: services/scheme_mapper.py
import json
from typing import Any


def first_value(*values: Any) -> Any:
    for value in values:
        if value not in (None, "", [], {}):
            return value
    return ""


def text_from(value: Any) -> str:
    if value in (None, ""):
        return ""
    if isinstance(value, str):
        return value.strip()
    if isinstance(value, list):
        return "; ".join(filter(None, (text_from(item) for item in value)))
    if isinstance(value, dict):
        preferred = first_value(
            value.get("en"),
            value.get("text"),
            value.get("value"),
            value.get("name"),
            value.get("label"),
        )
        return text_from(preferred) if preferred else json.dumps(value, ensure_ascii=True)
    return str(value)


def find_scheme_array(raw: Any) -> list[dict[str, Any]]:
    if isinstance(raw, list):
        return raw
    if not isinstance(raw, dict):
        return []

    candidates = [
        raw.get("data"),
        raw.get("schemes"),
        raw.get("results"),
        raw.get("response"),
        raw.get("payload", {}).get("schemes") if isinstance(raw.get("payload"), dict) else None,
    ]

    if isinstance(raw.get("data"), dict):
        candidates.extend([raw["data"].get("schemes"), raw["data"].get("results")])

    for candidate in candidates:
        if isinstance(candidate, list):
            return candidate
    return []


def map_scheme(scheme: dict[str, Any]) -> dict[str, Any]:
    name = first_value(scheme.get("scheme_name"), scheme.get("schemeName"), scheme.get("name"), scheme.get("title"))
    description = first_value(
        scheme.get("description"),
        scheme.get("schemeDescription"),
        scheme.get("briefDescription"),
        scheme.get("summary"),
    )
    benefits = first_value(scheme.get("benefits"), scheme.get("benefit"), scheme.get("benefitDetails"))
    eligibility = first_value(
        scheme.get("eligibility"),
        scheme.get("eligibilityCriteria"),
        scheme.get("eligibility_details"),
    )
    category = first_value(scheme.get("category"), scheme.get("schemeCategory"), scheme.get("categoryName"))
    level = first_value(scheme.get("level"), scheme.get("schemeLevel"), scheme.get("schemeType"))

    return {
        "id": text_from(first_value(scheme.get("id"), scheme.get("schemeId"), scheme.get("slug"), name)),
        "title": text_from(name) or "Untitled scheme",
        "description": text_from(description),
        "benefits": text_from(benefits),
        "eligibility": text_from(eligibility),
        "category": text_from(category) or "Uncategorized",
        "level": text_from(level) or "Scheme",
        "raw": scheme,
    }


def map_schemes(raw: Any) -> list[dict[str, Any]]:
    return [map_scheme(scheme) for scheme in find_scheme_array(raw) if isinstance(scheme, dict)]


def filter_schemes(schemes: list[dict[str, Any]], query: str, category: str) -> list[dict[str, Any]]:
    query = str(query or "").strip().lower()
    filtered = []

    for scheme in schemes:
        matches_category = category == "All categories" or scheme["category"] == category
        haystack = " ".join(str(scheme.get(key, "")) for key in ["title", "description", "benefits", "eligibility", "category"]).lower()
        matches_query = not query or query in haystack
        if matches_category and matches_query:
            filtered.append(scheme)

    return filtered
